h2o-algos/src/main/java/hex/optimization/OptimizationUtils.java

```diff
@@ -455,5 +459,104 @@ public GradientInfo ginfo() {
       return _ginfox;
     }
   }
+  
+  public static final class ExactLineSearch {
+    public final double _betaLS1 = 1e-4;
+    public final double _betaLS2 = 0.99;
+    public final double _lambdaLS = 2;
+    public double _alphal;
+    public double _alphar;
+    public double _alphai;
+    public double[] _direction;
+    public int _maxIteration = 20;  // limit LS steps to save time
+    public double[] _originalBeta;
+    public double[] _newBeta;
+    public GLM.GLMGradientInfo _ginfo;
+    public double _gradDireInner;
+
+    public ExactLineSearch(double[] betaCnd, ComputationState state) {
+      reset(betaCnd, state);
+    }
+    
+    public void reset(double[] betaCnd, ComputationState state) {
+      _direction = new double[betaCnd.length];      
+      ArrayUtils.subtract(betaCnd, state.beta(), _direction);
+      _ginfo = state.ginfo();
+      _originalBeta = state.beta();
+      _alphai = 1;
+      _alphal = 0;
+      _alphar = Double.POSITIVE_INFINITY;
+    }
+
+    /***
+     * Evaluate and make sure that step size alphi is not too big so that objective function is still decreasing.
+     */
+    public boolean evaluateFirstWolfe(GLM.GLMGradientInfo ginfo) {
+      double newObj = ginfo._objVal;
+      _gradDireInner = ArrayUtils.innerProduct(_ginfo._gradient, _direction);
+      double rhs = _ginfo._objVal+_alphai*_betaLS1*_gradDireInner;
+      return (newObj <= rhs);
+    }
+
+    /***
+     * Evaluate and make sure that step size alphi is not too small so that good
+     * progress is made in reducing the loss function.
+     */
+    public boolean evaluateSecondWolfe(GLM.GLMGradientInfo ginfo) {
+      double lhs = ArrayUtils.innerProduct(ginfo._gradient, _direction);
+      return lhs >= _betaLS2*_gradDireInner;
+    }
+
+    public boolean setAlphai(boolean firstWolfe, boolean secondWolfe) {
+      if (!firstWolfe) { // step is too long
+        _alphar = _alphai;
+        _alphai = 0.5*(_alphal+_alphar);
+        return true;
+      } else if (firstWolfe && !secondWolfe) { // step is too short
+        _alphal = _alphai;
+        if (_alphar < Double.POSITIVE_INFINITY)
+          _alphai = 0.5*(_alphal+_alphar);
+        else
+          _alphai = _lambdaLS * _alphai;
+        return true;
+      }
+      return false;
+    }
 
+    public boolean findAlpha(double[] lambdaEqual, double[] lambdaLessThan, ComputationState state,
+                             ConstrainedGLMUtils.LinearConstraints[] equalityConstraints, 
+                             ConstrainedGLMUtils.LinearConstraints[] lessThanEqualToConstraints,
+                             GLM.GLMGradientSolver ginfo) {
+      GLM.GLMGradientInfo newGrad;
+      assert ArrayUtils.innerProduct(_direction, _ginfo._gradient) <= 0 : "direction is not an descent direction!";
```

Comment:
It might be better to do:
```java
if (ArrayUtils.innerProduct(_direction, _ginfo._gradient) > 0)
    return false;
```

Running one of your tests (`pyunit_GH_6722_loose_only_linear_constraints_binomial.py`) I encountered this assertion error - the innerProduct was ~1e-7 and most of the values in gradient were around 1e-10 so I think it might have been a numerical issue.

This change would end up with logging "LS failed" and exiting the `fitIRLSMCS`.

Or if you don't like that solution a better solution could be some nice error message that would explain the user what went wrong.

Reply:
Changed per your suggestions.